scipy/spatial/transform/_rotation.pyx

```diff
@@ -115,10 +115,72 @@ cdef inline void _quat_canonical(double[:, :] q) noexcept:
     for ind in range(n):
         _quat_canonical_single(q[ind])
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline void _get_angles(
+    double[:] _angles, bint extrinsic, bint symmetric, bint sign,
+    double lamb, double a, double b, double c, double d):
+    
+        # intrinsic/extrinsic conversion helpers
+        cdef int angle_first, angle_third
+        if extrinsic:
+            angle_first = 0
+            angle_third = 2
+        else:
+            angle_first = 2
+            angle_third = 0
+
+        cdef double half_sum, half_diff
+        cdef int case
+    
+        # Step 2
+        # Compute second angle...
+        _angles[1] = 2 * atan2(hypot(c, d), hypot(a, b))
+
+        # ... and check if equal to is 0 or pi, causing a singularity
+        if abs(_angles[1]) <= 1e-7:
+            case = 1
+        elif abs(_angles[1] - <double>pi) <= 1e-7:
+            case = 2
+        else:
+            case = 0 # normal case
+
+        # Step 3
+        # compute first and third angles, according to case
+        half_sum = atan2(b, a)
+        half_diff = atan2(d, c)
+        
+        if case == 0:  # no singularities
+            _angles[angle_first] = half_sum - half_diff
+            _angles[angle_third] = half_sum + half_diff
+        
+        else:  # any degenerate case
+            _angles[2] = 0
+            if case == 1:
+                _angles[0] = 2 * half_sum
+            else:
+                _angles[0] = 2 * half_diff * (-1 if extrinsic else 1)
+                
+        # for Tait-Bryan angles
+        if not symmetric:
+            _angles[angle_third] *= sign
+            _angles[1] -= lamb
+
+        for idx in range(3):
+            if _angles[idx] < -pi:
+                _angles[idx] += 2 * pi
+            elif _angles[idx] > pi:
+                _angles[idx] -= 2 * pi
+
+        if case != 0:
+            warnings.warn("Gimbal lock detected. Setting third angle to zero "
+                          "since it is not possible to uniquely determine "
+                          "all angles.")
```

Comment:
Think the stack level might need to be set here so the warning shows up from the right place

Reply:
How do I do that?
I don't know if that's relevant but there are multiple tests expecting the warning:
```python
    with pytest.warns(UserWarning, match="Gimbal lock"):
```
That I think would fail if there was a problem with the warning

Reply:
I understand it as `stacklevel=2` (or even 3?) argument can be added such that the warning is issued from the caller function. 

It can make it easier for a user to interpret the problem.

@j-bowhay am I correct?

Reply:
Exactly, I just did not have time to look at the call stack to work out what number it should be set to. It should be straightforward just by playing around to see what it should be set to

Reply:
Apparently the default is `stacklevel=1`: https://python.readthedocs.io/en/latest/library/warnings.html#warnings.warn
Should I make it `stacklevel=2`? Or should I just put the warning outside of this internal function?

Reply:
`stacklevel=3` should put the warning into `as_euler/as_davenport` which I think will be the best.